### DISCUSSION TITLE
BB3 3B memory module querying - object has no attribute

### DISCUSSION THREAD

**Discussion by krisstud** [facebookresearch/ParlAI#5012] [Q&A] [answered]
According to the BB3 3B documentation, it should be possible to work with memories through the agent object [here](https://github.com/facebookresearch/ParlAI/blob/fae3e39c249d202c1c588fb3d9f9cbf6d017606a/projects/bb3/agents/README.md#bb3-3b-5) using .set_memory, .get_memory, get_available_memories() etc. 

This probably some silly configuration error, but i can't find it. Using the following code i get `AttributeError: 'ComboFidGoldDocumentAgent' object has no attribute 'get_available_memories'`. May this be related to loading the seeker agent instead of r2c2_bb3 agent?

The code
```
from parlai.core.agents import create_agent

opt_dict = {'model':'projects.bb3.agents.r2c2_bb3_agent:BB3SubSearchAgent','interactive_mode': True, 'init_opt' : 'gen/r2c2_bb3.opt', 'model_file':'/path/ParlAI/data/models/bb3/bb3_3B/model'}
bb3_agent = create_agent(opt_dict) 
bb3_agent.get_available_memories()

```

I'm aware that i can pass personas using observation like so:
```
agent.observe({'text': 'your persona: You are a female\nyour persona: your name is jenny', 'episode_done': False})
agent.observe({'text': 'Whats your name?', 'episode_done': False})
response = agent.act()
```
Background information is that if the model is shown context longer than the truncation limit (even using several  agent.observation that in themselves do not exceed it) it does not seem to store long-term memories. I would like to query the memory directly to confirm this. I saw a while back that another user had similar problems with long-term storage of memories in BB2 (can't find the link) and solved it using 'memory_key':'full_text', which isn't available in BB3 afaik.

**Comment by klshuster** [comment, marked as answer]
if you're initializing directly, you may need to manually specify all of the opt values in the [`gen/r2c2_bb3.opt`](https://github.com/facebookresearch/ParlAI/blob/main/parlai/opt_presets/gen/r2c2_bb3.opt). this includes using a different `--model` argument

**Reply by krisstud** [reply]
> if you're initializing directly, you may need to manually specify all of the opt values in the [`gen/r2c2_bb3.opt`](https://github.com/facebookresearch/ParlAI/blob/main/parlai/opt_presets/gen/r2c2_bb3.opt). this includes using a different `--model` argument

Copying the opts you linked to into a JSON string variable, and adding these opts to it works: `bb3_agent = create_agent_from_model_file('zoo:bb3/bb3_3B/model', opt_dict_own)`:
```
  "interactive_mode": "True",
  "force_skip_retrieval": "True", # https://github.com/facebookresearch/ParlAI/issues/4830 - disable search
```
Querying memory, i see that passing a custom context by agent.observe does not lead to memory generation. From projects/bb3/agents/README.md it is clear that one can interact with the memory module directly by for example manually calling `I am a big fan of the New York Yankees __generate-memory__` in context. Is there an easier way to make the model automatically perform memory generation from previous conversation turns than appending memory decisions and performing memory generation for each of them?

**Reply by klshuster** [reply]
With BB3 that is the best way to do it

**Reply by krisstud** [reply]
I ended up creating a custom safe_interactive script that also takes the previous context, calling it using --safety none, creating a custom batch_act that takes also takes the label from the context as input and uses that instead of reply generation, a custom batch_act_decision that never searches, and finally looping over the context using observe, custom batch_act and self_observe. This "hack" works and generates relevant memory decisions and memories before entering parley, but.. 

i run out of cuda memory when the total memory count exceeds ~15-16. I see from nvidia-smi that each memory added can use up to 1-2GB of GPU memory. What to do? My memory generation related opts are as follows:

```
"memory_generator_control_token": "__generate-memory__",
  "mgm_beam_block_ngram": 3,
  "mgm_beam_min_length": 5,
  "mgm_beam_size": 3,
  "mgm_inference": "beam",
  "mgm_history_size": 1,
  "mgm_model": "projects.bb3.agents.r2c2_bb3_agent:BB3SubSearchAgent",
```

decision:
 ```
 "mdm_beam_block_ngram": -1,
  "mdm_beam_min_length": 1,
  "mdm_beam_size": 1,
  "mdm_history_size": -1,
  "mdm_inference": "greedy",
  "mdm_model": "projects.bb3.agents.r2c2_bb3_agent:BB3SubSearchAgent",
  "memory_decision": "compute",
  "memory_decision_control_token": "__is-memory-required__",
  "memory_decision_do_access_reply": "__do-access-memory__",
  "memory_decision_dont_access_reply": "__do-not-access-memory__",
  "memory_decision_use_memories": true,
```

**Reply by klshuster** [reply]
You may be running out of memory due to the context length growing. If you type `[DONE]` after each interaction it will reset the conversation history

**Reply by krisstud** [reply]
> You may be running out of memory due to the context length growing. If you type `[DONE]` after each interaction it will reset the conversation history

Appended that to every turn. Same result. Can i somehow set a limit (maximum) on the number of memories?

**Reply by klshuster** [reply]
Unfortunately not in the current code
